game-core/src/main/java/games/strategy/triplea/settings/SelectionComponentFactory.java

```diff
@@ -412,28 +416,77 @@ public void reset() {
   }
 
   static SelectionComponent<JComponent> textField(final ClientSetting<String> clientSetting) {
-    return new AlwaysValidInputSelectionComponent() {
-      final JTextField textField = new JTextField(clientSetting.getValue().orElse(""), 20);
+    return textField(clientSetting, value -> value, encodedValue -> encodedValue, "");
+  }
+
+  static <T> SelectionComponent<JComponent> textField(
+      final ClientSetting<T> clientSetting,
+      final ThrowingFunction<T, String, ValueEncodingException> encodeValue,
+      final ThrowingFunction<String, T, ValueEncodingException> decodeValue,
+      final String validValueDescription) {
+    return new SelectionComponent<JComponent>() {
+      private final JTextField textField = new JTextField(encode(clientSetting::getValue), 20);
+
+      private String encode(final Supplier<Optional<T>> valueSupplier) {
+        return valueSupplier.get()
+            .map(value -> {
+              try {
+                return encodeValue.apply(value);
+              } catch (final ValueEncodingException e) {
+                log.log(
+                    Level.FINE,
+                    String.format("Failed to encode value '%s' from client setting '%s'", value, clientSetting),
+                    e);
+                return null;
+              }
+            })
+            .orElse("");
+      }
 
       @Override
       public JComponent getUiComponent() {
         return textField;
       }
 
+      @Override
+      public boolean isValid() {
+        final String encodedValue = textField.getText();
+        return encodedValue.isEmpty() || decode(encodedValue).isPresent();
```

Comment:
It was intentional to keep the `isEmpty()` check separate from `decode()`--I could have easily had that method return an empty `Optional` in the case the encoded value is an empty string.  However, when I address the `FIXME` below, I'm pretty sure these tests will need to be separate (see original implementation in first commit).  If not, I will hide the `isEmpty()` check appropriately at that time.

(This reasoning also applies to the same separation in `save()` below.)